File: android/src/main/kotlin/net/nend/nendplugin/AdBridger.kt
package net.nend.nendplugin

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.Locale

abstract class AdBridger {

    abstract val methodChannel: MethodChannel

    open fun onMethodCall(call: MethodCall, result: MethodChannel.Result) = result.notImplemented()

    fun MethodChannel.invokeListenerEvent(event: CallbackName, args: Map<String, Any>?) {
        this.invokeMethod(
            event.toString().decapitalize(Locale.ROOT),
            args
        )
    }

    protected fun targetMethod(method: String): MethodName {
        return MethodName.valueOf(method.capitalize(Locale.ROOT))
    }

    enum class MethodName {
        LoadAd,
        ReleaseAd,
        ShowAd,
        HideAd,
        // Common definitions are above here.

        Resume,
        Pause,

        IsReady,
        MediationName,
        UserId,
        UserFeature,
        AddFallbackFullboard,
        EnableAutoReload,
        DismissAd,
        MuteStartPlaying,

        // Specific definitions are above here.
        InitAd,
        SetLogLevel,
    }

    enum class CallbackName {
        OnLoaded,
        OnFailedToLoad,
        OnAdClicked,
        OnInformationClicked,
        // Common definitions are above here.

        OnFailedToPlay,
        OnShown,
        OnClosed,
        OnStarted,
        OnStopped,
        OnCompleted,
        OnRewarded,
        OnFailedToShow,
        OnDetectedVideoType,
        // Specific definitions are above here.
    }

    companion object {
        const val KEY_RESULT = "result"
    }
}

File: ios/Classes/AdBridger.swift
//
//  AdBridger.swift
//  nend_plugin
//
//

import UIKit
import Flutter

class AdBridger: NSObject {

    var channel: FlutterMethodChannel!

    var rootViewController: UIViewController? {
        return UIApplication.shared.keyWindow?.rootViewController
    }

    func adUnit(argument: Any?) -> AdUnit {
        let jsonData = try! JSONSerialization.data(withJSONObject: argument!, options: [])
        return try! JSONDecoder().decode(AdUnit.self, from: jsonData)
    }

    func invokeMethod(_ method: CallbackName, arguments: [String: Any?]?) {
        self.channel.invokeMethod(method.rawValue, arguments: arguments)
    }

    func targetMethod(_ method: String) -> MethodName {
        return MethodName(rawValue: method) ?? MethodName.notFound
    }

    enum MethodName: String {
        case loadAd,
        releaseAd,
        showAd,
        hideAd,
        // Common definitions are above here.

        resume,
        pause,

        isReady,
        mediationName,
        addFallbackFullboard,
        enableAutoReload,
        dismissAd,
        muteStartPlaying,

        // Specific definitions are above here.
        initAd,
        setLogLevel,
        notFound
    }

    enum CallbackName: String {
        case onLoaded,
        onFailedToLoad,
        onAdClicked,
        onInformationClicked,
        // Common definitions are above here.

        onFailedToPlay,
        onShown,
        onClosed,
        onStarted,
        onStopped,
        onCompleted,
        onRewarded,
        onFailedToShow,
        onReceiveAd,
        onDetectedVideoType,
        // Specific definitions are above here.
        notFound
    }
}
